### ISSUE TITLE
Not able to use an existing FF profile using GeckoDriver 20, FF 58/59, WIN10

### ISSUE THREAD

**Issue by javierlopez111** [mozilla/geckodriver#1242]
## System

Selenium-java code is as follows,

String profileDir = "C:/Users/jlopez/AppData/Roaming/Mozilla/Firefox/Profiles/ddtkw80l.profile5902";
FirefoxProfile ffprofile = new FirefoxProfile();

FirefoxOptions options = new FirefoxOptions();
options.setBinary("C:\Program Files\Mozilla Firefox\firefox.exe");
options.addArguments("-profile", profileDir);

System.setProperty("webdriver.gecko.driver","C:\selenium\drivers\geckodriver.exe");

capabilities = DesiredCapabilities.firefox();
capabilities.setCapability("moz:firefoxOptions", options);

driver = new RemoteWebDriver(new URL("http://" + hubHost + ":4567/wd/hub"), capabilities);

What am I doing wrong?? FF opens up but it does not go to the URL I specify.
If I comment out options.addArguments("-profile", profileDir); , it works ok, but with a new profile every time that does not have flash installed which is installed on the profile I have previously created.

* Version: 
geckodriver 20
* Platform: 
Windows 10
* Firefox: 
FF 58
* Selenium: 
Selenium standalone 3.10

## Testcase

Attempting to use an existing FF profile.
The code is below,
String profileDir = "C:/Users/jlopez/AppData/Roaming/Mozilla/Firefox/Profiles/ddtkw80l.profile5902";
FirefoxProfile ffprofile = new FirefoxProfile();

FirefoxOptions options = new FirefoxOptions();
options.setBinary("C:\Program Files\Mozilla Firefox\firefox.exe");
options.addArguments("-profile", profileDir);

System.setProperty("webdriver.gecko.driver","C:\selenium\drivers\geckodriver.exe");

capabilities = DesiredCapabilities.firefox();
capabilities.setCapability("moz:firefoxOptions", options);

driver = new RemoteWebDriver(new URL("http://" + hubHost + ":4567/wd/hub"), capabilities);

FF opens up but it does not go to the URL I specify.
If I comment out options.addArguments("-profile", profileDir); , it works ok, but with a new profile every time that does not have flash installed which is installed on the profile I have previously created.

**Comment by whimboo** [comment]
You are using a remote webdriver session which executes the tests on a remote but not the local machine, right? Given that you create the profile locally, it cannot be used on the remote machine unless you transfer the profile to the target machine. 

Otherwise use the `profile` key of FirefoxOptions to specify a base64 encoded profile. That should be the more elegant solution in your case.

**Comment by javierlopez111** [comment]
Yup, I understand the profile object would not exist in my local machine from which I am running my tests.  My bad.

I am now doing the following which I've tried in the past, but no cigar...

FirefoxOptions options = new FirefoxOptions();
                //options.addArguments("firefox_profile", "profile59022");
                options.addPreference("firefox_profile", "profile59022");
                options.setBinary("C:\\Program Files\\Mozilla Firefox\\firefox.exe");

I've tried passing it as argument, but that did not work.
I don't see a self describing  method under FirefoxOptions to set the profile, so I tried as a preference.
In either of the two above cases, FF is opened with a new profile.  I've included some logging.

Question: how do I make sure the profile is base64 encoded?? Is it just making sure I am using a 64bit FF version??

Thanks a bunch for your time and help.
Javier

11:34:33.602 INFO [ActiveSessionFactory.apply] - Capabilities are: Capabilities {acceptInsecureCerts: true, browserName: firefox, firefox_binary: C:/Program Files/Mozilla Fi..., moz:firefoxOptions: {args: [], binary: C:/Program Files/Mozilla Fi..., prefs: {firefox_profile: profile59022}}, version: }
11:34:33.602 INFO [ActiveSessionFactory.lambda$apply$11] - Matched factory org.openqa.selenium.remote.server.ServicedSession$Factory (provider: org.openqa.selenium.firefox.GeckoDriverService)
1523298873868   geckodriver     INFO    geckodriver 0.20.0
1523298873883   geckodriver     INFO    Listening on 127.0.0.1:18637
1523298874420   mozrunner::runner       INFO    Running command: "C:/Program Files/Mozilla Firefox/firefox.exe" "-marionette" "-profile" "C:\\Users\\jlopez\\AppData\\Local\\Temp\\rust_mozprofile.kHqQNx1n7yer"
1523298875520   Marionette      INFO    Enabled via --marionette
1523298876958   Marionette      INFO    Listening on port 52785
1523298877616   Marionette      WARN    TLS certificate errors will be ignored for this session
11:34:37.673 INFO [ProtocolHandshake.createSession] - Detected dialect: W3C
11:34:37.673 INFO [RemoteSession$Factory.lambda$performHandshake$0] - Started new session 5700b9d3-dd79-45b5-85a8-c55aba94b753 (org.openqa.selenium.firefox.GeckoDriverService)

**Comment by whimboo** [comment]
What you want is to do the following:

1. Create a profile and customize it so that it fits your needs
2. Create a zip archive from the profile directory
3. Base64 encode that zip archive
4. Use `options.profile` with that base64 encoded profile data

That way the profile will be transmitted to the remote machine, and used as template. Does that work for you?

**Comment by javierlopez111** [comment]
Do you have a suggestion of where to log a bug against??

**Comment by whimboo** [comment]
@javierlopez111 please make the fixes to your code first. As of now you clearly use it the wrong way. See the first part of my last comment.

**Comment by javierlopez111** [comment]
whimboo,

options.setProfile ONLY takes FirefoxProfile which in turn takes a directory location.  I am feeding the base64 zip file of the profile folder.  What exactly am I doing wrong??  I am sorry, I am  not a full fledge developer, so might need to explain my issue in simpler terms?  I really appreciate your help.

**Comment by whimboo** [comment]
Ok, so I had a look at https://seleniumhq.github.io/selenium/docs/api/java/org/openqa/selenium/firefox/FirefoxOptions.html and it seems that the Java bindings do not support that yet. They only allow a profile path set via `args`, but not `profile`. You want to file an issue here: https://github.com/SeleniumHQ/selenium/issues/

**Comment by javierlopez111** [comment]
I have also tried,
options.addArguments("-profile", "m7fmwi17.profile59022");

It opens the browser but it does nothing after that.  I believe others have run into this in the past, and it's still not fixed? Wherever the issue is...

https://github.com/mozilla/geckodriver/issues/1058

I will try FF59 as suggested on the last update.

**Comment by javierlopez111** [comment]
Thanks so much for your help !!

**Comment by whimboo** [comment]
> options.addArguments("-profile", "m7fmwi17.profile59022");
No, passing in a profile via the argument requires a path to the local profile as I mentioned above. This will not work.
